Accessibility issues addressing UX-1573

clrux opened the pull request:
This work is related to [UX-1573] (https://openedx.atlassian.net/browse/UX-1573). It provides a way to send focus to the content area after clicking an item in the sequence/unit navigation.

Currently when one activates a link in the sequence/unit navigation, focus is lost and sent back to the top of the page, forcing keyboard users to navigate through all of the elements until he or she reaches the content area. This work sends focus directly to the content area upon navigation link activation.

---

@cptvitamin When you get time, would you mind reviewing?

---

@frrrances @talbs Would one of you review the FED here?

cptvitamin commented on lms/templates/seq_module.html:
```diff
@@ -44,7 +44,7 @@
+  <div id="seq_content" tabindex="-1"></div>
 
   <nav class="sequence-bottom">
     <ul aria-label="${_('Section Navigation')}" class="sequence-nav-buttons">
```
this aria-label needs to go on the nav element 1 line up.  The word "Navigation" is not necessary as the SR will say that when it announces the type of landmark.

cptvitamin commented on common/lib/xmodule/xmodule/css/sequence/display.scss:
```diff
@@ -427,6 +427,13 @@ nav.sequence-bottom {
+
+  &:focus,
+  &:active {
+    outline: none;
```
This doesn't seem to be working? I am getting a visible focus ring in FF, Chrome and Safari on Mac

clrux replied:
Hmm, oddly this IS working for me. Let's catch up in person?

talbs commented on common/lib/xmodule/xmodule/css/sequence/display.scss:
```diff
@@ -438,6 +445,5 @@ nav.sequence-bottom {
     outline: 0;
-  }
-*/
+*/  }
```
@clrux, looks like @cptvitamin did you a solid with https://github.com/edx/edx-platform/commit/e75fe342dfca9f56da0d98b10914dc047400e01d. Can you double-check that this change 1) solves the compilation problem and 2) preserves proper Sass nesting and indentation structure?

clrux replied:
All fixed up, thanks @cptvitamin. We might want to revisit (read: cleanup) this file in the future.

talbs commented:
@clrux, I'm happy to review this. Can you get a quick sandbox spun up so I can see this focus behavior in the wild?

clrux commented on common/lib/xmodule/xmodule/css/sequence/display.scss:
```diff
@@ -120,17 +116,6 @@ nav.sequence-nav {
-          //   font-size: 12px;
-          //   color: #aaa;
-          // }
-
```
I removed a bunch of stuff that was commented out...

talbs replied:
Thanks for cleaning this up.

clrux commented on common/lib/xmodule/xmodule/css/sequence/display.scss:
```diff
@@ -235,7 +219,8 @@ nav.sequence-nav {
 
-        &:hover, &:focus {
+        &:hover,
+        &:focus {
```
Personal NITs for readability.

talbs replied:
If we're going to do this type of small NIT work, I'd rather us do it a clean up-based PR that does it across the application. Mind saving this change for that and reverting?

clrux commented:
@cptvitamin I've updated the ARIA stuff. As for the focusing, I wasn't seeing it (but this was also after I fixed some Sass compilation errors). Maybe give it another go? Also, I have no idea why my devstack spun up with those errors... wondering if I was on another branch?

@talbs Thanks bud. I'm getting a sandbox up for you now...

talbs commented on common/lib/xmodule/xmodule/css/sequence/display.scss:
```diff
@@ -268,7 +253,8 @@ nav.sequence-nav {
       z-index: 1;
 
-      &.prev, &.next {
+      &.prev,
```
If we're going to do this type of small NIT work, I'd rather us do it a clean up-based PR that does it across the application. Mind saving this change for that and reverting?

talbs commented on common/lib/xmodule/xmodule/css/sequence/display.scss:
```diff
@@ -429,15 +407,14 @@ nav.sequence-bottom {
+}
+
+.xmodule_VerticalModule div.vert-mod > div ul.sequence-nav-buttons {
+  list-style: none !important;
```
Do you really need the ``!important`` here? if so, I'd consider moving this to the ``_shame.scss`` partial with a note about where/why its done.

clrux replied:
Weird, I didn't add this...

talbs replied:
Perhaps @cptvitamin did.

cptvitamin replied:
Nothing is *that* !important.  Tom Giannattasio did this in 2012: https://github.com/edx/edx-platform/commit/2c827c46f9fd1f2baee0d9408d34bd80533fdedb

talbs commented:
@clrux and @cptvitamin, I'm seeing some not so great jumping behavior on the sandbox when navigating between units. Here's a low-quality screengrab - http://cl.ly/image/2S2F2e3C092G.

I've talked with @cptvitamin about this briefly already, he has some ideas.

talbs commented on common/lib/xmodule/xmodule/css/sequence/display.scss:
```diff
@@ -429,15 +405,10 @@ nav.sequence-bottom {
-    outline: 0;
+  &:focus,
+  &:active {
+    outline: none;
```
Any chance we can make this a Sass extend that can be consistently applied in all of the cases where we don't want an outline? Conversely, it'd be great to have an outline extend we use when we do want consistent show visual focus.

talbs replied:
Feel free to not do this now too - its not essential.

clrux commented on common/lib/xmodule/xmodule/js/src/sequence/display.coffee:
```diff
@@ -114,6 +114,8 @@ class @Sequence
       sequence_links = @content_container.find('a.seqnav')
       sequence_links.click @goto
+
+      @content_container.focus();
```
@talbs @cptvitamin Sending focus to the content container causes the jumping. Taking Mark's advice and adding an empty container just above it to which we'll send focus instead.

talbs commented on lms/templates/seq_module.html:
```diff
@@ -44,10 +44,11 @@
   </div>
   % endfor
-  <div id="seq_content" role="tabpanel"></div>
+  <div id="sr-is-focusable" tabindex="-1"></div>
```
NIT: should we separate this class into two classes: ``sr`` and ``is-focusable``?

clrux replied:
@talbs I'd considered this. However @cptvitamin said that having something `.sr`d would hide it and negate the focusable qualities. I opted for an `id` in this case... but a class might work better - only not `.sr`. Thoughts from either of you on a proper name?

cptvitamin replied:
@talbs @clrux, my concern with using sr is that it would be offscreen, and we *may* get a weird focus indicator that goes offscreen when someone clicks on a tab.  Since it is not visible anyway (It's empty) we don't need to move it offscreen with sr class.  We need an ID so that we can programmatically identify it with javascript with focus().  I think something short and sweet should work i.e. id="sr-focus"  as in, this is just here to move focus for screen reader users.

talbs replied:
Cool, carry-on (with the original/current class name) then, gents :)

talbs commented:
@clrux, aside from the NIT comment I just made, this approach seems solid. Any chance you can update the sandbox with the current DOM-element based solution or any other approaches?

clrux commented:
test this please

talbs commented:
@clrux, took a spin through your updated sandbox. The behavior for sighted users looks solid and introduces no jumping of the page/viewport. I can also tab right to where we wanted the page to be. Nice work!

:+1: to :truck: to Mergeburgh from my perspective.

You'll still want to get @cptvitamin's thumbs as the other reviewer.

talbs commented:
Also, one other note - please squash your commits into 1 or very few logical commits (with messages that are clear)

cptvitamin commented on lms/templates/seq_module.html:
```diff
@@ -7,7 +7,7 @@
 
     <div class="sequence-list-wrapper">
-      <ol role="tablist" aria-label="${_('Section Navigation')}" id="sequence-list">
+      <ol role="tablist" aria-label="${_('Section')}" id="sequence-list">
```
The aria-label needs to be on the parent nav element.  Remove aria-label (can't have that on an ```<ol>```) and put it on the parent ```<nav>``` and change the label to "Unit" or "Sequence".  We are using "Section" already on the left/right arrows at the bottom of the page.

cptvitamin commented:
Once you fix the aria-label issue, I say merge this.  This will improve the UX for a SR user dramatically.

clrux commented:
@talbs @cptvitamin One more pass, please?

talbs commented:
My :+1: is still holding strong up in the air for you (as soon as tests pass).

cptvitamin commented on lms/templates/seq_module.html:
```diff
@@ -1,13 +1,13 @@
 <div id="sequence_${element_id}" class="sequence" data-id="${item_id}" data-position="${position}" data-ajax-url="${ajax_url}" >
-  <nav class="sequence-nav">
+  <nav class="sequence-nav" aria-label="${_('Unit')}" >
     <ul class="sequence-nav-buttons">
```
LOVE LOVE LOVE this, and so will users.

cptvitamin commented:
:+1:

clrux commented:
test this please
